storage/src/u4.rs

```diff
@@ -0,0 +1,341 @@
+// Copyright (C) 2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE.md for licensing terms.
+
+/// An error similar to [`std::num::TryFromIntError`] but able to be created
+/// within our crate.
+///
+/// The std error does not have a public constructor, and neither does ours.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, thiserror::Error)]
+#[non_exhaustive]
+#[error("out of range integral type conversion attempted")]
```

Comment:
The error message here is identical to that of `std::num::TryFromIntError`. I went back and forth about providing the original value, the invalid value, the index where the invalid value occurs, and some other possible error message states. But, I decided that what stdlib does is sufficient because if the caller wanted to include the invalid value, they have the value that was passed into the constructor because `u8` and `&[u8]` are both `Copy` and cheap to carry around after the call.